### PR TITLE
Add merged odometry to firmware_message_converter

### PR DESCRIPTION
**Pull request by Bitterisland6**
Added publishing of odometry obtained from merging wheel odometry with imu data.

### CODE REVIEW

File: leo_fw/src/firmware_message_converter.cpp [hunk 1]
```diff
@@ -28,6 +29,17 @@ static ros::Subscriber imu_sub;
 static ros::Publisher imu_pub;
 static bool imu_advertised = false;
 
+static ros::Publisher odom_merged_pub;
+static ros::Timer odom_merged_timer;
+static geometry_msgs::Point odom_merged_position;
+static float odom_merged_yaw;
+static bool odom_merged_advertised = false;
+static float velocity_linear_x = 0;
+static float velocity_angular_z = 0;
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Use double instead of float

File: leo_fw/src/firmware_message_converter.cpp [hunk 2]
```diff
@@ -28,6 +29,17 @@ static ros::Subscriber imu_sub;
...
+ros::ServiceClient odom_reset_client;
+ros::ServiceServer reset_odometry_service;
+static constexpr float PI = 3.141592653F;
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Put this constant at the top after includes, remove static as its not needed and change float to double.

File: leo_fw/src/firmware_message_converter.cpp [hunk 3]
```diff
@@ -117,6 +135,35 @@ void imu_callback(const leo_msgs::ImuPtr& msg) {
...
+
+  const float move_x = velocity_linear_x * std::cos(odom_merged_yaw);
+  const float move_y = velocity_linear_x * std::sin(odom_merged_yaw);
+
+  odom_merged_position.x += move_x * 0.01;
+  odom_merged_position.y += move_y * 0.01;
+
+  odom_merged_yaw += velocity_angular_z * 0.01;
+
+  if (odom_merged_yaw > 2.0F * PI)
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: When working with doubles, don't use F suffix as it indicates a float literal.

File: leo_fw/src/firmware_message_converter.cpp [hunk 4]
```diff
@@ -28,6 +29,17 @@ static ros::Subscriber imu_sub;
...
 
+static ros::Publisher odom_merged_pub;
+static ros::Timer odom_merged_timer;
+static geometry_msgs::Point odom_merged_position;
+static float odom_merged_yaw;
+static bool odom_merged_advertised = false;
+static float velocity_linear_x = 0;
+static float velocity_angular_z = 0;
+ros::ServiceClient odom_reset_client;
+ros::ServiceServer reset_odometry_service;
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Make these static

**Review by bjsowa — CHANGES_REQUESTED** [review]

File: leo_fw/src/firmware_message_converter.cpp [hunk 5]
```diff
@@ -117,6 +136,35 @@ void imu_callback(const leo_msgs::ImuPtr& msg) {
...
+void merge_odometry_callback(const ros::TimerEvent& events) {
+  nav_msgs::Odometry merged_odom;
+  merged_odom.header.frame_id = odom_frame_id;
+  merged_odom.child_frame_id = robot_frame_id;
+  merged_odom.header.stamp = ros::Time::now();
+  merged_odom.twist.twist.linear.x = velocity_linear_x;
+  merged_odom.twist.twist.angular.z = velocity_angular_z;
+
+  const float move_x = velocity_linear_x * std::cos(odom_merged_yaw);
+  const float move_y = velocity_linear_x * std::sin(odom_merged_yaw);
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Change to double

File: leo_fw/src/firmware_message_converter.cpp [hunk 6]
```diff
@@ -117,6 +136,35 @@ void imu_callback(const leo_msgs::ImuPtr& msg) {
...
+
+  if (odom_merged_yaw > 2.0 * PI)
+    odom_merged_yaw -= 2.0 * PI;
+  else if (odom_merged_yaw < 0.0)
+    odom_merged_yaw += 2.0 * PI;
+
+  merged_odom.pose.pose.position.x = odom_merged_position.x;
+  merged_odom.pose.pose.position.y = odom_merged_position.y;
+  merged_odom.pose.pose.orientation.z = std::sin(odom_merged_yaw * 0.5F);
+  merged_odom.pose.pose.orientation.w = std::cos(odom_merged_yaw * 0.5F);
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Remove F suffix

File: leo_fw/src/firmware_message_converter.cpp [hunk 7]
```diff
@@ -173,6 +221,20 @@ bool set_imu_calibration_callback(leo_msgs::SetImuCalibrationRequest& req,
   return true;
 }
 
+bool reset_odometry_callback(std_srvs::TriggerRequest& req,
+                             std_srvs::TriggerResponse& res) {
+  odom_merged_position.x = 0.0F;
+  odom_merged_position.y = 0.0F;
+  odom_merged_yaw = 0.0F;
```
Comment (bjsowa) [inline leo_fw/src/firmware_message_converter.cpp]: Remove F suffix

**Review by bjsowa — CHANGES_REQUESTED** [review]

File: leo_fw/src/firmware_message_converter.cpp (hunk 7 above)
Comment (Bitterisland6) [reply leo_fw/src/firmware_message_converter.cpp]: Done

File: leo_fw/src/firmware_message_converter.cpp (hunk 6 above)
Comment (Bitterisland6) [reply leo_fw/src/firmware_message_converter.cpp]: Done

File: leo_fw/src/firmware_message_converter.cpp (hunk 5 above)
Comment (Bitterisland6) [reply leo_fw/src/firmware_message_converter.cpp]: Done

**Review by bjsowa — APPROVED** [review]

**Pull request merged**